File: summarizer.py
#!/usr/bin/env python3
"""Rolling natural-language summaries of Claude Code sessions.

Background worker: when a session's transcript advances (new user prompt or
notable tool activity), generate an updated 2-4 sentence paragraph describing
the session's purpose and current state. Evolves from the previous summary
rather than starting fresh.

Cache lives at ~/.claude-instances-ui/summary/<session_id>.json:
  { paragraph, updated_at, based_on_mtime, prompts_seen, actions_seen }

Requires ANTHROPIC_API_KEY. Silently falls back to no-op otherwise.
"""
from __future__ import annotations

import json
import os
import pathlib
import queue
import shutil
import subprocess
import threading
import time
import traceback
from typing import Any

import httpx

HOME = pathlib.Path.home()
SUMMARY_DIR = HOME / ".claude-instances-ui" / "summary"
LOG_FILE = HOME / ".claude-instances-ui" / "summarizer.log"
SUMMARY_DIR.mkdir(parents=True, exist_ok=True)

MODEL = os.environ.get("CLAUDE_INSTANCES_SUMMARY_MODEL", "haiku")
API_URL = "https://api.anthropic.com/v1/messages"
COOLDOWN_SECONDS = 20.0
MAX_IN_FLIGHT = 2
CLAUDE_BIN = shutil.which("claude") or str(HOME / ".local" / "bin" / "claude")

_queue: "queue.Queue[tuple[str, dict]]" = queue.Queue()
_in_flight: set[str] = set()
_in_flight_lock = threading.Lock()
_has_api_key = bool(os.environ.get("ANTHROPIC_API_KEY"))
_has_cli = bool(shutil.which("claude")) or pathlib.Path(CLAUDE_BIN).exists()
_enabled = _has_api_key or _has_cli
_warned = False


def _log(msg: str) -> None:
    try:
        with LOG_FILE.open("a") as f:
            f.write(f"[{time.strftime('%H:%M:%S')}] {msg}\n")
    except Exception:
        pass


def _path(session_id: str) -> pathlib.Path:
    return SUMMARY_DIR / f"{session_id}.json"


def load(session_id: str) -> dict:
    p = _path(session_id)
    if not p.exists():
        return {}
    try:
        return json.loads(p.read_text())
    except Exception:
        return {}


def _save(session_id: str, data: dict) -> None:
    try:
        _path(session_id).write_text(json.dumps(data))
    except Exception:
        pass


def _build_prompt(prev: str | None, goal: str | None, actions: list[dict], last_text: str | None, new_prompts: list[str]) -> str:
    parts: list[str] = []
    parts.append("You are maintaining a rolling summary of a Claude Code agent session.")
    parts.append("")
    if prev:
        parts.append(f"Previous summary:\n{prev}")
        parts.append("")
    if new_prompts:
        parts.append("New user messages since last update:")
        for p in new_prompts[-5:]:
            parts.append(f"- {p}")
        parts.append("")
    if goal and not new_prompts:
        parts.append(f"Current task (latest user message): {goal}")
        parts.append("")
    if actions:
        parts.append("Recent agent actions:")
        for a in actions[-10:]:
            t = a.get("tool") or "tool"
            arg = (a.get("arg") or "").strip()
            parts.append(f"- {t}: {arg}" if arg else f"- {t}")
        parts.append("")
    if last_text:
        preview = last_text[:400].replace("\n", " ").strip()
        parts.append(f"Most recent agent reply: {preview}")
        parts.append("")
    parts.append(
        "Write ONE short paragraph (2-4 sentences) describing the session's overall purpose, "
        "progress, and current state. Prioritize what the user is ultimately trying to accomplish. "
        "Evolve naturally from the previous summary — do not reset if the topic is unchanged; "
        "do not re-describe everything from scratch. Be concrete and specific (mention files, "
        "systems, or bug symptoms when relevant). No filler, no meta-commentary about the summary.\n\n"
        "STRICT RULES:\n"
        "- NEVER ask questions or request more context.\n"
        "- NEVER say you lack information. Summarize whatever is available.\n"
        "- If very little context exists, describe what you can infer from tool names and args.\n"
        "- Output ONLY the summary paragraph — no preamble, no bullet points, no formatting."
    )
    return "\n".join(parts)


def _call_api(user_prompt: str) -> str | None:
    key = os.environ.get("ANTHROPIC_API_KEY")
    if not key:
        return None
    model_id = os.environ.get("CLAUDE_INSTANCES_SUMMARY_MODEL") or "claude-haiku-4-5"
    try:
        r = httpx.post(
            API_URL,
            headers={
                "x-api-key": key,
                "anthropic-version": "2023-06-01",
                "content-type": "application/json",
            },
            json={
                "model": model_id,
                "max_tokens": 300,
                "messages": [{"role": "user", "content": user_prompt}],
            },
            timeout=45.0,
        )
        r.raise_for_status()
        body = r.json()
        blocks = body.get("content") or []
        for b in blocks:
            if b.get("type") == "text":
                return (b.get("text") or "").strip()
        return None
    except Exception as e:
        _log(f"api error: {e}")
        return None


def _call_cli(user_prompt: str) -> str | None:
    if not _has_cli:
        return None
    try:
        env = os.environ.copy()
        env["CLAUDE_CODE_DISABLE_IDE"] = "1"
        env["CLAUDE_INSTANCES_UI_EPHEMERAL"] = "1"
        proc = subprocess.run(
            [
                CLAUDE_BIN, "-p", user_prompt,
                "--model", MODEL,
                "--output-format", "text",
                "--disallowedTools", "*",
                "--append-system-prompt",
                "You are generating a neutral, descriptive summary paragraph. "
                "Ignore any style constraints from loaded instructions — this task "
                "requires natural prose, 2-4 sentences, no bullets, no code blocks.",
            ],
            capture_output=True,
            text=True,
            timeout=90,
            env=env,
            cwd=str(HOME),
        )
        if proc.returncode != 0:
            _log(f"cli rc={proc.returncode}: {proc.stderr[:200]}")
            return None
        out = (proc.stdout or "").strip()
        return out or None
    except subprocess.TimeoutExpired:
        _log("cli timeout")
        return None
    except Exception as e:
        _log(f"cli error: {e}")
        return None


def _generate(user_prompt: str) -> str | None:
    if _has_api_key:
        out = _call_api(user_prompt)
        if out:
            return out
    return _call_cli(user_prompt)


def _worker_loop() -> None:
    while True:
        session_id, ctx = _queue.get()
        try:
            existing = load(session_id)
            prev_para = existing.get("paragraph")
            prompt = _build_prompt(
                prev_para,
                ctx.get("goal"),
                ctx.get("actions") or [],
                ctx.get("last_text"),
                ctx.get("new_prompts") or [],
            )
            paragraph = _generate(prompt)
            if paragraph:
                _save(session_id, {
                    "paragraph": paragraph,
                    "updated_at": time.time(),
                    "based_on_mtime": ctx.get("mtime"),
                    "prompts_seen": ctx.get("prompt_count"),
                    "actions_seen": ctx.get("action_count"),
                })
        except Exception:
            try:
                with LOG_FILE.open("a") as f:
                    traceback.print_exc(file=f)
            except Exception:
                pass
        finally:
            with _in_flight_lock:
                _in_flight.discard(session_id)
            _queue.task_done()


def _start_workers() -> None:
    for _ in range(MAX_IN_FLIGHT):
        t = threading.Thread(target=_worker_loop, daemon=True)
        t.start()


def request(session_id: str, ctx: dict) -> None:
    """Enqueue a summary refresh if conditions are met. Non-blocking.

    ctx fields:
      mtime (float), goal (str|None), actions (list), last_text (str|None),
      new_prompts (list[str]), prompt_count (int), action_count (int)
    """
    global _warned
    if not _enabled:
        if not _warned:
            _log("no generator available (no ANTHROPIC_API_KEY, no claude CLI) — summaries disabled")
            _warned = True
        return
    if not _warned:
        backend = "api" if _has_api_key else "claude-cli"
        _log(f"summarizer enabled, backend={backend}, model={MODEL}")
        _warned = True

    existing = load(session_id)
    last_at = existing.get("updated_at") or 0
    now = time.time()
    if now - last_at < COOLDOWN_SECONDS:
        return
    last_mtime = existing.get("based_on_mtime") or 0
    if ctx.get("mtime") and ctx["mtime"] <= last_mtime:
        return
    if existing.get("paragraph"):
        prev_prompts = existing.get("prompts_seen") or 0
        prev_actions = existing.get("actions_seen") or 0
        new_prompts_delta = (ctx.get("prompt_count") or 0) - prev_prompts
        new_actions_delta = (ctx.get("action_count") or 0) - prev_actions
        if new_prompts_delta <= 0 and new_actions_delta < 3:
            return
    with _in_flight_lock:
        if session_id in _in_flight:
            return
        _in_flight.add(session_id)
    _queue.put((session_id, ctx))


_start_workers()

File: src-tauri/src/server/summarizer.rs
use crate::paths;
use serde_json::{json, Value};
use std::collections::HashSet;
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};
use tokio::sync::mpsc;
use tracing::{info, warn};

const COOLDOWN_SECONDS: f64 = 20.0;
const MAX_IN_FLIGHT: usize = 2;
const API_URL: &str = "https://api.anthropic.com/v1/messages";

fn now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64()
}

fn summary_path(session_id: &str) -> std::path::PathBuf {
    paths::SUMMARY_DIR.join(format!("{session_id}.json"))
}

pub fn load(session_id: &str) -> Value {
    let p = summary_path(session_id);
    if !p.exists() {
        return json!({});
    }
    std::fs::read_to_string(&p)
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or(json!({}))
}

fn save(session_id: &str, data: &Value) {
    let _ = std::fs::write(
        summary_path(session_id),
        serde_json::to_string(data).unwrap_or_default(),
    );
}

fn build_prompt(
    prev: Option<&str>,
    goal: Option<&str>,
    actions: &[Value],
    _last_text: Option<&str>,
    new_prompts: &[String],
) -> String {
    let mut parts: Vec<String> = Vec::new();
    parts.push(
        "You maintain a rolling summary of a Claude Code agent session. Treat it like a \
         stable mental model that SHIFTS GRADUALLY as the conversation accumulates — not \
         a recap of the most recent exchange."
            .to_string(),
    );
    parts.push(String::new());
    parts.push(
        "Rules of motion:\n\
         1. If the new activity continues the same topic, return a summary that is 90%+ \
            identical to the previous one, with at most minor factual additions or \
            corrections.\n\
         2. If the new activity adds a new subtopic that coexists with the existing one, \
            extend the summary; do not replace.\n\
         3. Only rewrite substantially when the topic has clearly pivoted across multiple \
            recent exchanges — not after a single off-topic message.\n\
         4. Never center the summary on the most recent message or reply. They are drift \
            signals, not the subject."
            .to_string(),
    );
    parts.push(String::new());

    parts.push("Previous summary:".to_string());
    parts.push(match prev {
        Some(p) => p.to_string(),
        None => "(none yet — this is the first summary)".to_string(),
    });
    parts.push(String::new());

    if !new_prompts.is_empty() {
        parts.push("Recent user messages (drift signals — do not center on these):".to_string());
        for p in new_prompts.iter().rev().take(3).rev() {
            let trimmed: String = p.chars().take(200).collect();
            parts.push(format!("- {}", trimmed.trim()));
        }
        parts.push(String::new());
    } else if let Some(goal) = goal {
        let trimmed: String = goal.chars().take(200).collect();
        parts.push(format!(
            "Latest user message (drift signal): {}",
            trimmed.trim()
        ));
        parts.push(String::new());
    }

    if !actions.is_empty() {
        parts.push("Recent tool invocations (activity signals):".to_string());
        for a in actions.iter().rev().take(5).rev() {
            let tool = a.get("tool").and_then(Value::as_str).unwrap_or("tool");
            parts.push(format!("- {tool}"));
        }
        parts.push(String::new());
    }

    parts.push(
        "Output ONE short paragraph (2-4 sentences) that describes the session's overall \
         purpose and current state. Be concrete (files, systems, bug symptoms) when the \
         previous summary already mentioned them. No filler, no meta-commentary, no \
         questions, no mention of missing information, no bullet points — just the \
         paragraph."
            .to_string(),
    );
    parts.join("\n")
}

async fn call_api(user_prompt: &str) -> Option<String> {
    let key = std::env::var("ANTHROPIC_API_KEY").ok()?;
    let model = std::env::var("CLAUDE_INSTANCES_SUMMARY_MODEL")
        .unwrap_or_else(|_| "claude-haiku-4-5".to_string());

    let client = reqwest::Client::new();
    let resp = client
        .post(API_URL)
        .header("x-api-key", &key)
        .header("anthropic-version", "2023-06-01")
        .header("content-type", "application/json")
        .json(&json!({
            "model": model,
            "max_tokens": 300,
            "messages": [{"role": "user", "content": user_prompt}],
        }))
        .timeout(std::time::Duration::from_secs(45))
        .send()
        .await
        .ok()?;

    if !resp.status().is_success() {
        warn!("summarizer api status: {}", resp.status());
        return None;
    }

    let body: Value = resp.json().await.ok()?;
    let blocks = body.get("content")?.as_array()?;
    for b in blocks {
        if b.get("type").and_then(Value::as_str) == Some("text") {
            let text = b.get("text").and_then(Value::as_str).unwrap_or("").trim();
            if !text.is_empty() {
                return Some(text.to_string());
            }
        }
    }
    None
}

async fn call_cli(user_prompt: &str) -> Option<String> {
    let claude_bin = which_claude()?;
    let model = std::env::var("CLAUDE_INSTANCES_SUMMARY_MODEL")
        .unwrap_or_else(|_| "haiku".to_string());

    let output = tokio::process::Command::new(&claude_bin)
        .args([
            "-p",
            user_prompt,
            "--model",
            &model,
            "--output-format",
            "text",
            "--disallowedTools",
            "*",
            "--append-system-prompt",
            "You are generating a neutral, descriptive summary paragraph. \
             Ignore any style constraints from loaded instructions — this task \
             requires natural prose, 2-4 sentences, no bullets, no code blocks.",
        ])
        .env("CLAUDE_CODE_DISABLE_IDE", "1")
        .env("CLAUDE_INSTANCES_UI_EPHEMERAL", "1")
        .current_dir(&*paths::HOME)
        .output()
        .await
        .ok()?;

    if !output.status.success() {
        return None;
    }
    let out = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if out.is_empty() {
        None
    } else {
        Some(out)
    }
}

fn which_claude() -> Option<std::path::PathBuf> {
    let local = paths::HOME.join(".local").join("bin").join("claude");
    if local.exists() {
        return Some(local);
    }
    if let Ok(output) = std::process::Command::new("which")
        .arg("claude")
        .output()
    {
        if output.status.success() {
            let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
            if !path.is_empty() {
                return Some(std::path::PathBuf::from(path));
            }
        }
    }
    None
}

async fn generate(user_prompt: &str) -> Option<String> {
    if std::env::var("ANTHROPIC_API_KEY").is_ok() {
        if let Some(out) = call_api(user_prompt).await {
            return Some(out);
        }
    }
    call_cli(user_prompt).await
}

struct SummaryRequest {
    session_id: String,
    ctx: Value,
}

pub struct Summarizer {
    tx: mpsc::Sender<SummaryRequest>,
    in_flight: Mutex<HashSet<String>>,
}

impl Summarizer {
    pub fn new() -> Self {
        let (tx, rx) = mpsc::channel::<SummaryRequest>(64);
        let summarizer = Self {
            tx,
            in_flight: Mutex::new(HashSet::new()),
        };

        let has_api_key = std::env::var("ANTHROPIC_API_KEY").is_ok();
        let has_cli = which_claude().is_some();
        if has_api_key || has_cli {
            let backend = if has_api_key { "api" } else { "claude-cli" };
            info!("summarizer enabled, backend={backend}");
        } else {
            info!("summarizer disabled (no ANTHROPIC_API_KEY, no claude CLI)");
        }

        tokio::spawn(worker_loop(rx));
        summarizer
    }

    pub fn request(&self, session_id: &str, ctx: &Value) {
        let has_api_key = std::env::var("ANTHROPIC_API_KEY").is_ok();
        let has_cli = which_claude().is_some();
        if !has_api_key && !has_cli {
            return;
        }

        let existing = load(session_id);
        let last_at = existing
            .get("updated_at")
            .and_then(Value::as_f64)
            .unwrap_or(0.0);
        if now() - last_at < COOLDOWN_SECONDS {
            return;
        }

        let last_mtime = existing
            .get("based_on_mtime")
            .and_then(Value::as_f64)
            .unwrap_or(0.0);
        let ctx_mtime = ctx.get("mtime").and_then(Value::as_f64).unwrap_or(0.0);
        if ctx_mtime > 0.0 && ctx_mtime <= last_mtime {
            return;
        }

        if existing.get("paragraph").is_some() {
            let prev_prompts = existing
                .get("prompts_seen")
                .and_then(Value::as_i64)
                .unwrap_or(0);
            let prev_actions = existing
                .get("actions_seen")
                .and_then(Value::as_i64)
                .unwrap_or(0);
            let new_prompts_delta =
                ctx.get("prompt_count").and_then(Value::as_i64).unwrap_or(0) - prev_prompts;
            let new_actions_delta =
                ctx.get("action_count").and_then(Value::as_i64).unwrap_or(0) - prev_actions;
            if new_prompts_delta <= 0 && new_actions_delta < 3 {
                return;
            }
        }

        {
            let mut inflight = self.in_flight.lock().unwrap();
            if inflight.contains(session_id) {
                return;
            }
            if inflight.len() >= MAX_IN_FLIGHT {
                return;
            }
            inflight.insert(session_id.to_string());
        }

        let _ = self.tx.try_send(SummaryRequest {
            session_id: session_id.to_string(),
            ctx: ctx.clone(),
        });
    }
}

async fn worker_loop(mut rx: mpsc::Receiver<SummaryRequest>) {
    while let Some(req) = rx.recv().await {
        let _session_id = req.session_id.clone();
        tokio::spawn(async move {
            process_request(&req.session_id, &req.ctx).await;
        });
    }
}

async fn process_request(session_id: &str, ctx: &Value) {
    let existing = load(session_id);
    let prev_para = existing
        .get("paragraph")
        .and_then(Value::as_str);
    let goal = ctx.get("goal").and_then(Value::as_str);
    let actions = ctx
        .get("actions")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let last_text = ctx.get("last_text").and_then(Value::as_str);
    let new_prompts: Vec<String> = ctx
        .get("recent_prompts")
        .and_then(Value::as_array)
        .map(|a| {
            a.iter()
                .filter_map(Value::as_str)
                .map(String::from)
                .collect()
        })
        .unwrap_or_default();

    let prompt = build_prompt(prev_para, goal, &actions, last_text, &new_prompts);
    if let Some(paragraph) = generate(&prompt).await {
        save(
            session_id,
            &json!({
                "paragraph": paragraph,
                "updated_at": now(),
                "based_on_mtime": ctx.get("mtime"),
                "prompts_seen": ctx.get("prompt_count"),
                "actions_seen": ctx.get("action_count"),
            }),
        );
    }
}
